### PR TITLE
Fix the bugs in RLP/Tx/PI circuit which are reported by Zellic & KALOS auditors

### PR DESCRIPTION
**Pull request by kunxian-xia**
### Description

This PR aims to fix all the bugs found by Zellic auditors in RLP & Tx & PI circuits.

### Issue Link

The bug report can be found [here](https://hackmd.io/4XuxkwHERa6AIPQ6tr2QAg).

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### Contents

1. [x] finding 3: https://github.com/scroll-tech/zkevm-circuits/pull/575
2. [x] finding 4: https://github.com/scroll-tech/zkevm-circuits/pull/576
3. [x] finding 5: https://github.com/scroll-tech/zkevm-circuits/pull/584
4. [x] finding 6: https://github.com/scroll-tech/zkevm-circuits/pull/586
5. [x] finding 7: https://github.com/scroll-tech/zkevm-circuits/pull/625
6. [x] finding 10: https://github.com/scroll-tech/zkevm-circuits/pull/578
7. [x] finding 11: https://github.com/scroll-tech/zkevm-circuits/pull/719
8. [x] finding 13: https://github.com/scroll-tech/zkevm-circuits/pull/579
9. [x] finding 14: https://github.com/scroll-tech/zkevm-circuits/pull/580
10. [x] finding 17: https://github.com/scroll-tech/zkevm-circuits/pull/602
11. [x] tx & pi: https://github.com/scroll-tech/zkevm-circuits/pull/612

### CODE REVIEW

**Comment by icemelon** [comment]
Could you keep 1 PR per finding?

**Comment by kunxian-xia** [comment]
What about 1 commit per finding? Because each fix can be done within about 10 lines of codes.

**Comment by icemelon** [comment]
No, still better to have one PR per finding. after merged, it'll become 1 commit per fix

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 1]
```diff
@@ -781,6 +808,28 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+        diffs.extend(byte_value_gte_0xb8.lt_chip.config().diff);
+        diffs.extend(byte_value_gte_0xc0.lt_chip.config().diff);
+        diffs.extend(byte_value_gte_0xf8.lt_chip.config().diff);
+        diffs.extend(byte_value_lte_0x80.lt_chip.config().diff);
+        diffs.extend(byte_value_lte_0xb8.lt_chip.config().diff);
+        diffs.extend(byte_value_lte_0xc0.lt_chip.config().diff);
+        diffs.extend(byte_value_lte_0xf8.lt_chip.config().diff);
+        diffs.extend(tidx_lte_tlength.lt_chip.config().diff);
+        diffs.extend(mlength_lte_0x20.lt_chip.config().diff);
+
```
Comment (sampritipanda) [inline zkevm-circuits/src/rlp_circuit_fsm.rs]: The new comparator `tlength_lte_mlength` might also need the `lt_chip.diff` columns to be range checked.

File: zkevm-circuits/src/rlp_circuit_fsm.rs [hunk 2]
```diff
@@ -572,13 +576,30 @@ impl<F: Field> RlpCircuitConfig<F> {
...
+            );
+
             cb.gate(meta.query_fixed(q_enabled, Rotation::cur()))
         });
 
         meta.lookup_any("byte value check", |meta| {
             let cond = and::expr([
                 meta.query_fixed(q_enabled, Rotation::cur()),
-                is_padding_in_dt.expr(Rotation::cur())(meta),
+                not::expr(is_padding_in_dt.expr(Rotation::cur())(meta)),
```
Comment (sampritipanda) [inline zkevm-circuits/src/rlp_circuit_fsm.rs:568]: We would suggest adding some negative tests to catch issues like this when changes are made to the circuit in the future.

File: zkevm-circuits/src/rlp_circuit_fsm.rs (hunk 1 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/rlp_circuit_fsm.rs]: fixed in https://github.com/scroll-tech/zkevm-circuits/pull/572/commits/baa4a7be0a68e8a3da63b6b2caeb0902d81ec90f.

File: zkevm-circuits/src/rlp_circuit_fsm.rs (hunk 2 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/rlp_circuit_fsm.rs:568]: Yeah, we can do that, but the `byte_value` field defined in `struct DataTable` has type `u8`. To add negative test like the above, we might need to change this to `u16` which however will skip the compiler check when we assign value larger than `2^8`. https://github.com/scroll-tech/zkevm-circuits/blob/25dd32aa316ec842ffe79bb8efe9f05f86edc33e/zkevm-circuits/src/witness/rlp_fsm.rs#L455-L472

**Comment by lispc** [comment]
ComparatorChip needs a u8 column now in develop.  Codes here can be refactored

```
impl<F: Field, const N_BYTES: usize> ComparatorChip<F, N_BYTES> {
    /// Configure the comparator chip. Returns the config.
    pub fn configure(
        meta: &mut ConstraintSystem<F>,
        q_enable: impl FnOnce(&mut VirtualCells<F>) -> Expression<F> + Clone,
        lhs: impl FnOnce(&mut VirtualCells<F>) -> Expression<F> + Clone,
        rhs: impl FnOnce(&mut VirtualCells<F>) -> Expression<F> + Clone,
        u8_table: TableColumn,
    ) -> ComparatorConfig<F, N_BYTES> {
```

**Comment by lispc** [comment]
great. we can mark the asana board to mark them all done

**Review by lispc — APPROVED** [review]
let us merge it after CI passes

**Pull request merged**